Extend multimodal/speech_llm with lhotse, t5 and bestow supports

zhehuaichen opened the pull request:
# What does this PR do ?

In multimodal/speech_llm, add lhotse dataloader support and two models, SALM-T5 and Bestow-GPT. Include example configs.


###  Main features under speech_llm
- Lhotse dataloader support for speech SFT in speech_llm
- SALM-style architecture with T5 LLM backbone
- Bestow-style architecture (cross-attention based) with GPT LLM backbone
### Minor edit in nlp collection: 
- megatron_base_model.py: handle the case tokenizer.type is not set
- megatron_lm_encoder_decoder_model.py: hanlde the case encoder_input is used 
- megatron_base_prompt_learning_model.py: group the llm init code under init_model function (follow the pattern from megatron_gpt_prompt_learning_model.py) so that it can be overwritten by subclass when needed
- megatron/utils.py: in gradient accumulation, handle the case where the batch size from dynamic bucketing is not divisible. This happens when using lhotse dataloader with batch_duration

**Collection**: [common,nlp,multimodal]


**PR Type**:
- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

stevehuang52 commented on examples/multimodal/speech_llm/conf/modular_audio_gpt_config_cross_llama_lhotse.yaml:
```diff
@@ -0,0 +1,361 @@
+# Copyright (c) 2024, NVIDIA CORPORATION.  All rights reserved.
```
could you put this file under `conf/bestow/*` ?

zhehuaichen replied:
Done

stevehuang52 commented on examples/multimodal/speech_llm/conf/modular_audio_gpt_config_llama_lhotse.yaml:
```diff
@@ -0,0 +1,349 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+name: megatron_audio_gpt_salm_lhotse
```
could you put this file under `conf/salm/*` ?

zhehuaichen replied:
Done

stevehuang52 commented on examples/multimodal/speech_llm/conf/modular_audio_t5_config.yaml:
```diff
@@ -0,0 +1,338 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+name: megatron_audio_t5_salm_lhotse
```
could you put this file under `conf/salm/*` ?

zhehuaichen replied:
Done

pzelasko commented on nemo/collections/multimodal/speech_llm/data/lhotse_dataset.py:
```diff
@@ -0,0 +1,493 @@
+    vectors = collate_vectors_lhotse(items, padding_value=padding_value)
+    if max_length > vectors.size(1):
+        vectors = torch.cat(
+            [vectors, padding_value * torch.ones(vectors.size(0), max_length - vectors.size(1), dtype=vectors.dtype)],
```
why do we need to enforce a static shape with padding for every example here?

zhehuaichen replied:
to be consistent with the behavior in mtron dataloader and AudioTextDataset in nemo/collections/multimodal/speech_llm/data/audio_text_dataset.py

pzelasko replied:
but that kills the purpose of doing bucketing, doesn't it? at least to some extent - as far as I can tell we only do this for transcripts, but keep audio shapes dynamic

zhehuaichen replied:
yes

pzelasko commented on nemo/collections/multimodal/speech_llm/data/lhotse_dataset.py:
```diff
@@ -0,0 +1,493 @@
+    return (n + m - 1) // m * m
+
+
+class TextProcessing:
```
This class needs more documentation on what is it doing, how to use its API, and what are the expected input and output formats. Also, it only has private methods right now, the main API method should be public (no underscore at the beginning).

I'd expect a docstring of kind: ```this class is used to convert X to Y. in order to do so, it performs A, B, C, and D. the expect format of X is .... the expected format of Y is ...```

since it's used to convert text to prompts to token ids, I'd like to see full documentation of the prompt template/schema

zhehuaichen replied:
Done

pzelasko commented on nemo/collections/multimodal/speech_llm/data/lhotse_dataset.py:
```diff
@@ -0,0 +1,493 @@
+        return processed_example
+
+
+def convert_canary_prompt_to_text(prompt, is_canary_tokens_augment):
```
I understand why this function is built the way it is but for future experiments let's try to move away from canary special token conversion and design a configurable prompting setup instead.

zhehuaichen replied:
acknowledge. Add a TODO

pzelasko commented on nemo/collections/multimodal/speech_llm/data/lhotse_dataset.py:
```diff
@@ -0,0 +1,493 @@
+    by Lhotse samplers instead.
+    """
+
+    def __init__(
```
init args are not documented

zhehuaichen replied:
Done

pzelasko commented on nemo/collections/multimodal/speech_llm/data/lhotse_dataset.py:
```diff
@@ -0,0 +1,493 @@
+        tokens_to_generate: int,
+        pad_to_max_length: bool,
+        max_seq_length: int,
+        noise_cuts: Optional = None,
```
this is a left-over from before lhotse nemo dataset was refactored; remove noise_cuts in this class.

zhehuaichen replied:
Done

pzelasko commented on nemo/collections/multimodal/speech_llm/data/build_dataset.py:
```diff
@@ -0,0 +1,233 @@
+                conf['manifest_filepath'] = cur_manifest_filepath
+                question_file_set = data_cfg.get('question_file_set', None)
+                if question_file_set is not None:
+                    conf['question_file_set'] = [question_file_set[dataset_idx]]
```
unless you add `question_file_set` to `LhotseDataLoadingConfig` it will be discarded at the beginning of `get_lhotse_dataloader_from_config`

zhehuaichen replied:
Remove this flag

pzelasko commented on nemo/collections/multimodal/speech_llm/data/build_dataset.py:
```diff
@@ -0,0 +1,233 @@
+        from nemo.collections.common.data.lhotse import get_lhotse_dataloader_from_config
+
+        # for eval, we need to create separate dataset so as to report splitted numbers
+        if data_cfg.get('is_tarred', False) or (is_eval == False and is_predict == False):
```
you shouldn't rely on the flag `is_tarred` for choosing lhotse dataloader, lhotse supports more formats than nemo json and nemo tar and auto-deduces `is_tarred`

zhehuaichen replied:
good catch! Done.

pzelasko commented on examples/multimodal/speech_llm/conf/modular_audio_gpt_config_llama_lhotse.yaml:
```diff
@@ -0,0 +1,349 @@
+  freeze_modality_adapter: False
+  load_audio_encoder: True
+
+  global_batch_size: 128
```
how do these settings of global/micro batch size work with lhotse dynamic batch sizes?

zhehuaichen replied:
added comment
```
## Legacy batch_size configuration
  # When used with lhotse, the batch composition is decided by dataloader configs
  # and batch size here is only used for deciding gradient accumulation.
  # gradient accumulation = global_batch_size / micro_batch_size / data_parallel_size
  # where data_parallel_size = num_nodes * num_gpus / TP_size
```

pzelasko commented on examples/multimodal/speech_llm/conf/modular_audio_gpt_config_llama_lhotse.yaml:
```diff
@@ -0,0 +1,349 @@
+        average: null # Average the metric over the dataset. Options: ['macro', 'micro']. Works only for 'F1', 'accuracy' etc. Refer to torchmetrics for metrics where this is supported.
+        num_classes: null
+
+    # test_ds:
```
remove?

zhehuaichen replied:
Done

stevehuang52 commented on nemo/collections/multimodal/speech_llm/parts/utils/data_utils.py:
```diff
@@ -155,3 +155,15 @@ def align_feat_seq_list(
     return new_seq_list, new_seq_len_list
+
+
+def to_cuda(inputs, non_blocking=True):
```
remove this function, use this one https://github.com/NVIDIA/NeMo/blob/main/nemo/collections/asr/parts/mixins/transcription.py#L70

zhehuaichen replied:
Done

stevehuang52 commented on nemo/collections/nlp/modules/common/megatron/utils.py:
```diff
@@ -417,8 +417,8 @@ def split_list(inputs, num_chunks):
     """
     Split a list into equal sized chunks
     """
+    # if len(inputs) % chunk_size != 0, round down the chunk size
```
remove this line if it's not used

zhehuaichen replied:
not used but try to provide extra explanation on what happens

stevehuang52 commented on nemo/collections/nlp/modules/common/megatron/utils.py:
```diff
@@ -442,8 +442,11 @@ def get_iterator_k_split(batch: Union[Dict, List[torch.Tensor]], num_microbatche
 
         # Split tensor items
         items = list(tensor_items.items())
-        assert items[0][1].shape[0] % num_microbatches == 0, "Issue with batch size configuration!"
```
need someone from NLP to review the change on removing the constraint `assert items[0][1].shape[0] % num_microbatches == 0`

zhehuaichen replied:
acknowledge

aklife97 replied:
is there a way we can separate the cases where we know the batch size is not divisible with cases where things should be divisible?

silently rounding off can cause silent errors which can cause accuracy regressions and be super difficult to figure out

aklife97 replied:
looks good, thanks!

stevehuang52 commented on nemo/collections/nlp/models/language_modeling/megatron_base_model.py:
```diff
@@ -246,12 +246,12 @@ def __init__(self, cfg: DictConfig, trainer: Trainer, no_lm_init=True):
         self.use_fsdp = cfg.get('fsdp', False)
 
     def setup_transformer_engine_tp_groups(self):
-        """ This should be called after model parallel groups have been initialized
```
please go though the changes and undo anything that's not necessary

zhehuaichen replied:
from offline discussion, keep it for now since it is automatically applied after pushing the code https://github.com/NVIDIA/NeMo/pull/9169/commits/3cc0432ccb9dcab9b1d9bb52d17ad912fc21fbdf

stevehuang52 reviewed (changes requested):
Thanks for the great work, please address the CodeQL issues and see the minor comments.

zhehuaichen commented on examples/multimodal/speech_llm/conf/modular_audio_gpt_config_llama_lhotse.yaml:
```diff
@@ -0,0 +1,349 @@
+      bucketing_batch_size: null
+      use_lhotse: True
+      duration_bins : [2,4,6,8,10,12,14,16,18]
+      lhotse:
```
Done

zhehuaichen commented on nemo/collections/multimodal/speech_llm/data/build_dataset.py:
```diff
@@ -0,0 +1,233 @@
+from nemo.utils import logging
+
+
+def build_salm_dataset(model_instance, data_cfg, is_train):
```
This is a function for creating dataset more conveniently. Try to name this as salm style since we already have salm in conf/. I imagine we will add build_.*_dataset for other styles in future

stevehuang52 replied:
The 'salm' in config is referring to a specific set of params we use in the SALM paper, not referring to a general arch. Same for 'bestow'. we should use more general names than model-specific names, just like we're trying to avoid having 'canary' in the code except for the prompt format

zhehuaichen replied:
DOne

zhehuaichen commented on nemo/collections/multimodal/speech_llm/data/lhotse_dataset.py:
```diff
@@ -0,0 +1,493 @@
+    return (n + m - 1) // m * m
+
+
+class TextProcessing:
```
I am not sure whether making lhotse dataloader part inherits nemo mtron dataloader is a good idea for future since we have decided to move away from nemo one at some point. Added docstring of the cls. Hope it helps.

stevehuang52 replied:
when have we decided to move away from nemo? If they share most of the code, we can move the TextProcessing class to another file like data_utils.py so that both versions import from the same file

zhehuaichen replied:
Done

zhehuaichen commented on nemo/collections/multimodal/speech_llm/data/lhotse_dataset.py:
```diff
@@ -0,0 +1,493 @@
+        random_context_prob: float = 0.0,
+        random_context_positive_percent: float = 0.1,
+    ):
+        from lhotse.dataset import AudioSamples, CutMix
```
Done

stevehuang52 replied:
it's still not moved to the top

zhehuaichen replied:
Done

zhehuaichen commented on nemo/collections/multimodal/speech_llm/data/lhotse_dataset.py:
```diff
@@ -0,0 +1,493 @@
+        self.random_context_prob = random_context_prob
+        self.random_context_positive_percent = random_context_positive_percent
+
+    def _inject_random_context_into_question(self, cut, random_context_num=8, random_context_positive_percent=0.1):
```
include it in the implementation and keep default to be disabled. i think it is good to include this part of the implementation as how it is done in SALM and we can improve it in future research

zhehuaichen replied:
Done

zhehuaichen commented on nemo/collections/multimodal/speech_llm/data/lhotse_dataset.py:
```diff
@@ -0,0 +1,493 @@
+        return processed_example
+
+
+def convert_canary_prompt_to_text(prompt, is_canary_tokens_augment):
```
include a TODO on moving away from canary special token conversion and designing a configurable prompting setup instead
add a pointer on format in the comment

zhehuaichen commented on nemo/collections/multimodal/speech_llm/data/lhotse_dataset.py:
```diff
@@ -0,0 +1,493 @@
+                elif hasattr(cut, "question"):
+                    pass
+                else:
+                    cut.question = self.question + ' ' + canary_text
```
good catch! add an assert in else with comments

zhehuaichen commented on nemo/collections/multimodal/speech_llm/data/lhotse_dataset.py:
```diff
@@ -0,0 +1,493 @@
+    return x
+
+
+def _build_loss_mask(processed_example: dict, answer_only_loss: bool = True):
```
Done

zhehuaichen commented on nemo/collections/multimodal/speech_llm/data/lhotse_dataset.py:
```diff
@@ -0,0 +1,493 @@
+
+def adjust_input_ids(item: dict) -> dict:
+    """Mimics the logic from nemo/collections/multimodal/data/audio_text_qa_dataset.py:131"""
+    item["input_ids"] = item.get("masked_input_ids", item["input_ids"])
```
Done

zhehuaichen commented on nemo/collections/multimodal/speech_llm/models/modular_models.py:
```diff
@@ -121,11 +129,14 @@ def setup_optimizer_param_groups(self):
-        if self.cfg.get('freeze_llm', True):
-            for param in self.model.parameters():
-                param.requires_grad = False
+        self.unfreeze()
```
try to include cross attention transformer module in the training when exists. make the behavior as when a new module is added, it is jointly trained by default

zhehuaichen commented on nemo/collections/multimodal/speech_llm/models/modular_models.py:
```diff
@@ -1517,8 +1442,6 @@ def maybe_build_test(self):
             logging.info('Building test datasets...')
             # Wrap this in a list since the general finetuning parent class supports multi-validation.
             self._test_ds = self._build_dataset(self.cfg.data.test_ds, is_train=False)
-            lengths = [len(x) for x in self._test_ds]
```
lhotse dataloader cannot return the length of the dataset

zhehuaichen commented on nemo/collections/multimodal/speech_llm/models/modular_models.py:
```diff
@@ -836,7 +758,7 @@ def restore_from_pretrained_models(
 
         base_model_cfg = MegatronGPTSFTModel.merge_cfg_with(cfg.model.restore_from_path, cfg)
-        audio_model, audio_model_cfg = cls.get_audio_encoder_models_and_configs(cfg)
+        audio_model, audio_model_cfg = cls.get_audio_encoder_models_and_configs(cfg.model)
```
try to be able to call get_audio_encoder_models_and_configs from setup_perception_modules

stevehuang52 replied:
it's a bit confusing to change the signature of `get_audio_encoder_models_and_configs`, since the `cfg` in other class methods like `load_adapters_for_inference`, `load_pretrained_audio_weights` and `load_audio_encoder_for_inference` refers to the input yaml config, not just the model config. It's better to change the `cfg` param of get_audio_encoder_models_and_configs to another name so as to avoid confusion

zhehuaichen replied:
reverted the change

zhehuaichen commented on nemo/collections/multimodal/speech_llm/models/modular_models.py:
```diff
@@ -88,15 +85,26 @@
+                else MultiAudioPerceptionModule(cfg=cfg.perception)
+            )
+
+        if self._cfg.get('tensor_model_parallel_size', 1) == 1:
```
still needs more works. there was an error when loading asr model config interacts with TP. Add a warning

zhehuaichen commented on nemo/collections/multimodal/speech_llm/modules/perception_modules.py:
```diff
@@ -429,3 +430,118 @@ def forward(
+        *args,
+        **kwargs,
+    ):
+        assert input_embeds.shape[-1] == encoder_states.shape[-1]
```
Done

stevehuang52 replied:
not done yet

zhehuaichen replied:
Added docstring for forward()

zhehuaichen commented on nemo/collections/multimodal/speech_llm/models/modular_models_t5.py:
```diff
@@ -0,0 +1,1385 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import copy
```
rename the new one to modular_t5_models but keep the previous one to avoid breaking anything

zhehuaichen commented on nemo/collections/multimodal/speech_llm/models/modular_models.py:
```diff
@@ -763,16 +684,16 @@ def get_speaker_model_and_config(cls, cfg):
 
     @classmethod
     def get_audio_encoder_models_and_configs(cls, cfg):
-        if 'encoders' in cfg.model.perception:
```
tested for non NGC. will test NGC before merging

pzelasko commented on examples/multimodal/speech_llm/conf/salm/modular_audio_t5_config.yaml:
```diff
@@ -0,0 +1,338 @@
+      # sample_alpha: 0.1
+      use_lhotse: True
+      duration_bins : [2,4,6,8,10,12,14,16,18]
+      lhotse:
```
found another config that's using outdated lhotse args

zhehuaichen replied:
Done

stevehuang52 commented on nemo/collections/multimodal/speech_llm/data/build_dataset.py:
```diff
@@ -0,0 +1,228 @@
+            prepend_to_exist_question=data_cfg.get('prepend_to_exist_question', None),
+            canary_tokens_augment_ratio=data_cfg.get('canary_tokens_augment_ratio', 0.0),
+            random_context_prob=data_cfg.get('random_context_prob', 0.0),
+            random_context_positive_percent=data_cfg.get('random_context_positive_percent', 0.1),
```
please remove random context

zhehuaichen replied:
Done

stevehuang52 commented on nemo/collections/multimodal/speech_llm/data/build_dataset.py:
```diff
@@ -0,0 +1,228 @@
+def build_salm_dataloader(dataset, data_cfg, consumed_samples=0, is_predict=False, is_eval=False):
+    """Buld dataloader given an input dataset."""
+    if data_cfg.get("use_lhotse"):
+        if is_eval == False and is_predict == False:
```
you can get rid of `is_predict` and `is_eval` by accessing the `self.trainer.state.fn` variable to see whether it's `fit`, `predict` or something else

zhehuaichen replied:
Done

stevehuang52 commented on nemo/collections/multimodal/speech_llm/data/lhotse_dataset.py:
```diff
@@ -0,0 +1,517 @@
+        canary_tokens_augment_ratio: Ratio of canary tokens to augment
+        random_context_prob: Probability of injecting random context
+        random_context_positive_percent: Percentage of positive random context
+    """
```
1. please make it more general, instead of canary-specific
2. please remove random context fields

stevehuang52 replied:
This class is more like a `LhotseCanaryAudioQuestionAnswerDataset` as it depends on a lot of Canary-specific stuffs.....I think we should make the class more general

stevehuang52 commented on nemo/collections/multimodal/speech_llm/models/modular_t5_models.py:
```diff
@@ -0,0 +1,1385 @@
+        The list of microbatches is then piped through the pipeline using megatron-core fwd/bwd functions.
+        """
+        # # Initialize userbuffer communicators.
+        # if self.initialize_ub:
```
please remove commented out code

zhehuaichen replied:
done

stevehuang52 commented on nemo/collections/multimodal/speech_llm/modules/perception_modules.py:
```diff
@@ -429,3 +430,118 @@ def forward(
+        *args,
+        **kwargs,
+    ):
+        assert input_embeds.shape[-1] == encoder_states.shape[-1]
```
not done

stevehuang52 commented on nemo/collections/multimodal/speech_llm/modules/perception_modules.py:
```diff
@@ -429,3 +430,131 @@ def forward(
+
+
+class ProjectTransformerCrossAttention(NeuralModule, Exportable):
+    """Transformer module for cross-attention between speech and text embeddings.
```
why not add an optional MLP to the `TransformerCrossAttention` class and make the default as nn.Identity() such that we can avoid redundant code?

stevehuang52 commented on nemo/collections/nlp/modules/common/megatron/utils.py:
```diff
@@ -417,8 +417,8 @@ def split_list(inputs, num_chunks):
     """
+    # if len(inputs) % chunk_size != 0, round down the chunk size
     chunk_size = len(inputs) // num_chunks
-    assert len(inputs) % chunk_size == 0, "Issue with batch size configuration!"
```
need to ask someone from NLP to make sure they're fine with removing these asserts

zhehuaichen replied:
acknowledge. will do

stevehuang52 reviewed (changes requested):
Thanks for the great work, just a few comments. Also please remember to address the CodeQL issues. My main concerns are:
1. We need to remove random context and canary-specific dataloading, since they were only used in our internal experiments but not useful for external users. We added them for fast prototyping but they are not good designs.
2. We should avoid using `salm` and `bestow` in the dataloader function names, same reason why we don't name the model class as "SALM" or "BESTOW"
3. Please ask someone from NLP to see if they're okay with removing the assertions
4. Like @titu1994 mentioned in the previous [PR](https://github.com/NVIDIA/NeMo/pull/8741), we want to have flexible fields to use as "context", but it seems that the dataloader is only using `question` as they key for input text. It's better to keep it aligned with the non-lhotse dataset config, since it will be strange to suddenly find that the `input_key` in dataset config doesn't work as expected.
5. And some small comments.

zhehuaichen commented on nemo/collections/multimodal/speech_llm/data/lhotse_dataset.py:
```diff
@@ -0,0 +1,517 @@
+                cut.question = context + cut.question
+            self.random_context = current_words
+
+    def __getitem__(self, cuts) -> dict[str, torch.Tensor | list[str] | dict]:
```
it's loaded by lhotse  dataloader. have removed all the canary specific stuffs as discussed

zhehuaichen commented:
> Thanks for the great work, just a few comments. Also please remember to address the CodeQL issues. My main concerns are:
> 
> 1. We need to remove random context and canary-specific dataloading, since they were only used in our internal experiments but not useful for external users. We added them for fast prototyping but they are not good designs.
> 2. We should avoid using `salm` and `bestow` in the dataloader function names, same reason why we don't name the model class as "SALM" or "BESTOW"
> 3. Please ask someone from NLP to see if they're okay with removing the assertions
> 4. Like @titu1994 mentioned in the previous [PR](https://github.com/NVIDIA/NeMo/pull/8741), we want to have flexible fields to use as "context", but it seems that the dataloader is only using `question` as they key for input text. It's better to keep it aligned with the non-lhotse dataset config, since it will be strange to suddenly find that the `input_key` in dataset config doesn't work as expected.
> 5. And some small comments.

1. remove canary-specific parts and use lhotse dataloader support in getting "context" field. This is done by either reading the specified "context_key" field from the manifest or use the "default_context_key" field to extract from input_cfg in lhotse. An example input_cfg is as below so as to reuse existing canary manifests when the user wants
```
- input_cfg:
  - manifest_filepath: manifest1.json
    type: nemo
    weight: 2.0
    tags:
      default_context: "pls transcribe the audio"
  - manifest_filepath: manifest2.json
    type: nemo
    weight: 1.0
    tags:
      default_context: "pls translate English audio to German"
  type: group
  weight: 0.4
``` 

Remove random context.

2. Done
3. Will do after getting approval from speech_llm side
4. Done
5. Fixed

stevehuang52 reviewed (commented):
LGTM, please address the CodeQL comments (mainly remove unused imports) and I'll approve

stevehuang52 reviewed (approved):
LGTM, thanks

titu1994 commented on nemo/collections/multimodal/speech_llm/models/modular_models.py:
```diff
@@ -1561,3 +1475,75 @@ def list_available_models(cls) -> Optional[PretrainedModelInfo]:
         results.append(model)
         return results
+
+
```
Maybe put in another file, otherwise as we start getting more variants this file will explode in size

zhehuaichen replied:
Acknowledged. Try to put all similar gpt based speech_llm in modular_models.py for now. We can split to different categories when this list grows

titu1994 commented on nemo/collections/multimodal/speech_llm/models/modular_t5_models.py:
```diff
@@ -0,0 +1,1366 @@
+from nemo.core.classes.mixins import adapter_mixins
+from nemo.utils import AppState, logging, model_utils
+
+try:
```
Note: apex is being removed from NeMo - avoid using it and instead use mcores internals

zhehuaichen replied:
follow MegatronGPTSFTModel and MegatronT5PromptLearningModel that still uses it

titu1994 replied:
Right but apex will be removed in the main 2.0 release. It's fine in the PR but prepare to remove it in a month or so

titu1994 commented on nemo/collections/multimodal/speech_llm/models/modular_t5_models.py:
```diff
@@ -0,0 +1,1366 @@
+                    )
+                filename_log_key = self._determine_log_key(data_cfg, dataloader_idx, None, mode)
+                output_dir = data_cfg.get("output_dir", "./")
+                self.write_predictions_to_file(
```
Don't you need to put a rank guard ?

zhehuaichen replied:
follow MegatronGPTSFTModel to do the writing only on rank=0. i dont think it is necessary

titu1994 commented on nemo/collections/multimodal/speech_llm/models/modular_t5_models.py:
```diff
@@ -0,0 +1,1366 @@
+            raise ValueError("T5 does not support both distributed adam and mcore distributed data parallel.")
+
+
+class DecoderTextPromptModularizedAudioT5Model(ModularizedAudioT5Model):
```
Maybe  put into another file

zhehuaichen replied:
try to keep all the similar design T5 based speech_llm in this modular_t5_models.py file

titu1994 commented on nemo/collections/multimodal/speech_llm/modules/common/audio_text_generation_utils.py:
```diff
@@ -98,6 +98,7 @@ def send_generate_info(
         repetition_penalty,
         min_tokens_to_generate,
     ]
+    input_info.append(0)
```
?

zhehuaichen replied:
good catch! removed

titu1994 commented on nemo/collections/multimodal/speech_llm/modules/modality_adapters.py:
```diff
@@ -132,3 +132,15 @@ def forward(self, audio_signal, length=None):
+class IdentityConnectors(NeuralModule, Exportable, AccessMixin):
+    def __init__(
+        self,
+        *args,
```
Unless it's necessary, remove *args and **,kwargs

zhehuaichen replied:
this part is necessary since to keep the same set of params when initializing the different types of connectors

titu1994 replied:
Sounds reasonable

titu1994 commented on nemo/collections/multimodal/speech_llm/modules/perception_modules.py:
```diff
@@ -429,3 +430,78 @@ def forward(
+        input_lengths,
+        decoder_mems_list=None,
+        return_mems=False,
+        *args,
```
Why is *args **kwargs necessary ?

zhehuaichen replied:
removed

titu1994 reviewed (commented):
Overall, the PR looks ok. I would have liked a much leaner codebase without so much boilerplate code for Megatron tp pp etc, but it's unavoidable. 

My main comment is that most of the functions of the two new classes are undocumented, which needs to be fixed but not in this current pr.

zhehuaichen commented on nemo/collections/multimodal/speech_llm/models/modular_models.py:
```diff
@@ -1561,3 +1475,75 @@ def list_available_models(cls) -> Optional[PretrainedModelInfo]:
+        )
+        if 'audio_ratio' in audio_batch:
+            audio_ratio = audio_batch['audio_ratio'][..., None, None]
+            encoder_input = encoder_input * audio_ratio + input_embeds * (1 - audio_ratio)
```
Good catch! Does not work for inference. Need to call separate speech inference and text inference methods. We will work on this for joint speech SFT and text SFT. Add a TODO here

titu1994 replied:
Sounds good

zhehuaichen commented on examples/multimodal/speech_llm/conf/bestow/modular_audio_gpt_config_cross_llama_lhotse.yaml:
```diff
@@ -0,0 +1,329 @@
+      num_buckets : 30
+      buffer_size : 10000
+      shuffle_buffer_size : 10000
+      duration_bins: [2.92,3.474,3.924,4.335,4.728,5.11,5.487,5.872,6.288,6.696,7.128,7.62,8.208,8.934,9.883,10.56,11.22,11.88,12.51,13.05,13.59,14.13,14.64,15.17875,15.81,16.54,17.37,18.241,19.18]
```
Removed

zhehuaichen commented on examples/multimodal/speech_llm/conf/bestow/modular_audio_gpt_config_cross_llama_lhotse.yaml:
```diff
@@ -0,0 +1,329 @@
+  logger: False # logger provided by exp_manager
+  enable_checkpointing: False
+  use_distributed_sampler: False
+  max_epochs: 9999
```
acknowledged

titu1994 reviewed (dismissed):
Thanks for addressing the comments, lgtm !
